updater.go

```diff
@@ -50,8 +56,159 @@ func NewUpdater(source UpdateSource, config Config, log logging.Logger) *Updater
+	}
+
+	// Prompt for update
+	updateAction, err := u.promptForUpdateAction(ctx, *update, options)
+	if err != nil {
+		return update, err
+	}
+	switch updateAction {
+	case UpdateActionApply, UpdateActionAuto:
+		// Continue
+	case UpdateActionSnooze:
+		return update, cancelError("Snoozed update")
+	case UpdateActionCancel:
+		return update, cancelError("Canceled by user")
+	case UpdateActionError:
+		return update, fmt.Errorf("Error in update prompt: %s", err)
+	}
+
+	// Linux updates don't have assets so it's ok to prompt for update above before
+	// we check for nil asset
+	if update.Asset == nil || update.Asset.URL == "" {
+		u.log.Info("No update asset to apply")
+		return update, nil
+	}
+	if update.Asset.LocalPath == "" {
+		return update, fmt.Errorf("No local asset path")
+	}
+
+	// Ensure asset still exists in case it got pulled while we were waiting in
+	// the prompt.
+	exists, err := util.URLExists(update.Asset.URL, time.Minute, u.log)
+	if err != nil {
+		return update, err
+	}
+	if !exists {
+		return update, fmt.Errorf("Asset no longer exists: %s", update.Asset.URL)
+	}
+
+	u.log.Infof("Verify asset: %s", update.Asset.LocalPath)
+	if err := ctx.Verify(*update); err != nil {
+		return update, err
+	}
+
+	if err := ctx.BeforeApply(*update); err != nil {
+		return update, err
+	}
+
+	if err := u.platformApplyUpdate(*update, options); err != nil {
+		return update, err
+	}
+
+	if err := ctx.AfterApply(*update); err != nil {
+		return update, err
+	}
+
+	if err := ctx.Restart(); err != nil {
+		return update, err
+	}
+
+	return update, nil
+}
+
+// downloadAsset will download the update to a temporary path (if not cached),
+// check the digest, and set the LocalPath property on the asset.
+func (u *Updater) downloadAsset(asset *Asset, options UpdateOptions) error {
+	if asset == nil {
+		return fmt.Errorf("No asset to download")
+	}
+	downloadOptions := util.DownloadURLOptions{
+		Digest:        asset.Digest,
+		RequireDigest: true,
+		UseETag:       true,
+		Log:           u.log,
+	}
+
+	downloadPath := util.TempPath("", asset.Name+".")
+	if err := util.DownloadURL(asset.URL, downloadPath, downloadOptions); err != nil {
+		return err
+	}
+
+	asset.LocalPath = downloadPath
+	return nil
+}
+
+// checkForUpdate checks a update source (like a remote API) for an update.
+// It may set an InstallID, if the server tells us to.
+func (u *Updater) checkForUpdate(ctx Context, options UpdateOptions) (*Update, error) {
+	u.log.Infof("Checking for update, current version is %s", options.Version)
+	u.log.Infof("Using updater source: %s", u.source.Description())
+	u.log.Debugf("Using options: %#v", options)
+
+	update, findErr := u.source.FindUpdate(options)
+	if findErr != nil {
+		return nil, findErr
+	}
+	if update == nil {
+		return nil, nil
+	}
+
+	// Save InstallID if we received one
+	if update.InstallID != "" {
+		if err := u.config.SetInstallID(update.InstallID); err != nil {
+			u.log.Warningf("Error saving install ID: %s", err)
```

Comment:
I have the `Error() string` method returning the internal type in the description.